rfd/0199-teleport-key-agent.md

```diff
@@ -0,0 +1,514 @@
+---
+authors: Brian Joerger (bjoerger@goteleport.com)
+state: draft
+---
+
+# RFD 199 - Teleport Key Agent
+
+## Required Approvers
+
+* Engineering: @rosstimothy
+* Product: @xinding33 || @klizhentas
+
+## What
+
+Teleport Key Agent handles private key storage and operations for multiple
+local Teleport clients.
+
+## Why
+
+Teleport Key Agent will enable multiple use cases, including:
+
+* Providing a shared key state necessary for specific features
+  * Delegating Hardware Key touch/pin prompts to Teleport Connect for better UX
+  * [Caching hardware key pin](./0198-hardware-key-pin-caching.md) between
+    multiple Teleport clients
+  * Syncing Teleport Connect's front end login state with `tsh`
+  * Sharing login state stored in [memory](#memory-key-storage) without exposing private keys to disk
+* Providing concurrent access to key storage with limiting properties
+  * Example: Hardware Key Support holds an exclusive connection to the Hardware
+    Key, preventing concurrent access from different clients
+
+Note: the primary short-term goal of this RFD is to enable UX improvements in
+Hardware Key PIN support, while the other use cases above may or may not be
+addressed in the future. These extra features have been marked below as
+*future work*.
+
+## Details
+
+The client running the Teleport Key Agent will be referred to as the "agent"
+client. Any client interfacing with the agent will be called a "dependent"
+client.
+
+### UX
+
+In order to utilize the Teleport Key Agent, users will need to launch an agent
+client and login. See [running the agent](#running-the-agent)
+for more details on running the agent.
+
+Dependent clients will interface with the agent for some or all login information,
+depending on the [agent mode](#modes).
+
+Dependent clients should retrieve login information seamlessly from the agent
+so that there is no UX degradation when compared to a client solely using file
+storage.
+
+#### Running the Agent
+
+##### Teleport Connect
+
+Teleport Connect is an ideal runner for an agent client since it:
```

Comment:
If the short-term goal is to cache hardware key PIN across multiple tsh operations, how naive/possible would it be to start some kind of an agent in the background whenever a PIN is needed?

I suppose the advantage of having an agent in Connect is that it can cache the PIN for all sorts of operations, no matter if they come from standalone tsh invocations or some operations within Connect itself.

Reply:
IIUC that's essentially what we came up with at the start - have the Teleport client spawn a background PIV agent to hold the PIV key and PIN and kill it after the timeout (e.g. 15 minutes). However this isn't actually simpler, it just adds a new background process to manage. We still need to introduce an API for signing with the PIV key between the agent and the client.